csharp/extractor/Semmle.Extraction/Context.cs

```diff
@@ -43,28 +43,81 @@ public SemanticModel Model(SyntaxNode node)
         int NewId() => TrapWriter.IdCounter++;
 
         /// <summary>
-        /// Gets the cached label for the given entity, or creates a new
-        /// (cached) label if it hasn't already been created. The label
-        /// is set on the supplied <paramref name="entity"/> object.
+        /// Creates a new entity using the factory.
         /// </summary>
-        /// <returns><code>true</code> iff the label already existed.</returns>
-        public bool GetOrAddCachedLabel(ICachedEntity entity)
+        /// <param name="factory">The entity factory.</param>
+        /// <param name="init">The initializer for the entity.</param>
+        /// <returns>The new/existing entity.</returns>
+        public Entity CreateEntity<Type, Entity>(ICachedEntityFactory<Type, Entity> factory, Type init) where Entity : ICachedEntity
         {
-            var id = GetId(entity);
-            if (id == null)
-                throw new InternalError("Attempt to create a null entity for {0}", entity.GetType());
+            return init == null ? CreateEntity2(factory, init) : CreateNonNullEntity(factory, init);
+        }
 
-            Label existingLabel;
-            if (labelCache.TryGetValue(id, out existingLabel))
+        /// <summary>
+        /// Creates a new entity using the factory.
+        /// Uses a different cache to <see cref="CreateEntity{Type, Entity}(ICachedEntityFactory{Type, Entity}, Type)"/>,
+        /// and can store null values.
+        /// </summary>
+        /// <param name="factory">The entity factory.</param>
+        /// <param name="init">The initializer for the entity.</param>
+        /// <returns>The new/existing entity.</returns>
+        public Entity CreateEntity2<Type, Entity>(ICachedEntityFactory<Type, Entity> factory, Type type) where Entity : ICachedEntity
+        {
+            using (StackGuard)
             {
-                entity.Label = existingLabel;
-                return true;
+                var entity = factory.Create(this, type);
+
+                if(entityLabelCache.TryGetValue(entity, out var label))
+                {
+                    entity.Label = label;
+                    return entity;
+                }
+                else
+                {
+                    var id = entity.Id;
+                    label = new Label(NewId());
+                    entity.Label = label;
+                    entityLabelCache[entity] = label;
+                    DefineLabel(label, id);
+                    if (entity.NeedsPopulation)
+                        Populate(null, entity);
```

Comment:
Why always `null` argument, and not `init as ISymbol`?

Reply:
Looks like a bug....